File: gate-domain/src/main/java/gate/domain/ticket/Ticket.java
package gate.domain.ticket;

import java.time.Instant;
import java.util.List;

/**
 * Minimal ticket entity (执行文档 D4: kept as the aggregation unit for cost telemetry in P4).
 *
 * <p>{@code executor*} / {@code reviewer*} are nullable in P1 — the orchestration layer fills the
 * executor coordinates, and P1's reviewer is the {@code manual} provider.
 */
public record Ticket(
        String ticketNo,
        String title,
        String targetRef,
        String clonePath,
        String executorProviderId,
        String executorModel,
        String reviewerProviderId,
        String reviewerModel,
        TicketStage stage,
        Instant createdAt,
        Instant updatedAt,
        // P4 cost telemetry (nullable — usually NULL because this project does not spawn the agent)
        Long execTokenTotal,
        String execTokenSource,
        // S4: optional AgentConfig bound at creation time (nullable)
        String agentConfigId,
        // V5: queue priority P0..P3 (nullable = unset) and project ownership
        String priority,
        String projectId,
        // V8: editable work item content (nullable text + normalized labels)
        String description,
        String note,
        List<String> labels,
        // V19 快速模式: the project's permanent super ticket (clone_path IS the workspace,
        // never closes, bypasses the gate pipeline). Existing rows default to false.
        boolean isSuper) {

    /** Valid priority levels, ordered most urgent first. */
    public static final List<String> PRIORITIES = List.of("P0", "P1", "P2", "P3");
    public static final int MAX_LABELS = 20;
    public static final int MAX_LABEL_LENGTH = 32;

    /**
     * Backward-compatible constructor for callers that don't have P4 cost data (P1/P2 paths).
     * Passes null for both cost fields and the agent config.
     */
    public Ticket(String ticketNo, String title, String targetRef, String clonePath,
                  String executorProviderId, String executorModel,
                  String reviewerProviderId, String reviewerModel,
                  TicketStage stage, Instant createdAt, Instant updatedAt) {
        this(ticketNo, title, targetRef, clonePath, executorProviderId, executorModel,
                reviewerProviderId, reviewerModel, stage, createdAt, updatedAt, null, null, null,
                null, null, null, null, List.of());
    }

    /**
     * Backward-compatible constructor for P4 cost data without an agent config.
     */
    public Ticket(String ticketNo, String title, String targetRef, String clonePath,
                  String executorProviderId, String executorModel,
                  String reviewerProviderId, String reviewerModel,
                  TicketStage stage, Instant createdAt, Instant updatedAt,
                  Long execTokenTotal, String execTokenSource) {
        this(ticketNo, title, targetRef, clonePath, executorProviderId, executorModel,
                reviewerProviderId, reviewerModel, stage, createdAt, updatedAt,
                execTokenTotal, execTokenSource, null, null, null, null, null, List.of());
    }

    /**
     * Backward-compatible constructor for S4 (agent config, no V5 metadata).
     */
    public Ticket(String ticketNo, String title, String targetRef, String clonePath,
                  String executorProviderId, String executorModel,
                  String reviewerProviderId, String reviewerModel,
                  TicketStage stage, Instant createdAt, Instant updatedAt,
                  Long execTokenTotal, String execTokenSource, String agentConfigId) {
        this(ticketNo, title, targetRef, clonePath, executorProviderId, executorModel,
                reviewerProviderId, reviewerModel, stage, createdAt, updatedAt,
                execTokenTotal, execTokenSource, agentConfigId, null, null, null, null, List.of());
    }

    /**
     * Backward-compatible constructor for V5 metadata without editable work item content.
     */
    public Ticket(String ticketNo, String title, String targetRef, String clonePath,
                  String executorProviderId, String executorModel,
                  String reviewerProviderId, String reviewerModel,
                  TicketStage stage, Instant createdAt, Instant updatedAt,
                  Long execTokenTotal, String execTokenSource, String agentConfigId,
                  String priority, String projectId) {
        this(ticketNo, title, targetRef, clonePath, executorProviderId, executorModel,
                reviewerProviderId, reviewerModel, stage, createdAt, updatedAt,
                execTokenTotal, execTokenSource, agentConfigId, null, null, null, null, List.of());
    }

    /**
     * Backward-compatible constructor for the pre-V19 shape (no super-ticket flag): every
     * ticket materialized before V19 is a regular gate-pipeline work item.
     */
    public Ticket(String ticketNo, String title, String targetRef, String clonePath,
                  String executorProviderId, String executorModel,
                  String reviewerProviderId, String reviewerModel,
                  TicketStage stage, Instant createdAt, Instant updatedAt,
                  Long execTokenTotal, String execTokenSource, String agentConfigId,
                  String priority, String projectId,
                  String description, String note, List<String> labels) {
        this(ticketNo, title, targetRef, clonePath, executorProviderId, executorModel,
                reviewerProviderId, reviewerModel, stage, createdAt, updatedAt,
                execTokenTotal, execTokenSource, agentConfigId, priority, projectId,
                description, note, labels, false);
    }

    public Ticket {
        if (ticketNo == null || ticketNo.isBlank()) {
            throw new IllegalArgumentException("ticketNo must not be blank");
        }
        if (targetRef == null || targetRef.isBlank()) {
            throw new IllegalArgumentException("targetRef must not be blank");
        }
        if (stage == null) {
            throw new IllegalArgumentException("stage must not be null");
        }
        if (priority != null && !PRIORITIES.contains(priority)) {
            throw new IllegalArgumentException("priority must be one of " + PRIORITIES + " or null");
        }
        labels = labels == null ? List.of() : List.copyOf(labels);
        if (labels.size() > MAX_LABELS) {
            throw new IllegalArgumentException("at most " + MAX_LABELS + " labels");
        }
        for (String label : labels) {
            if (label == null || label.isBlank()) {
                throw new IllegalArgumentException("label must not be blank");
            }
            if (label.length() > MAX_LABEL_LENGTH) {
                throw new IllegalArgumentException(
                        "label longer than " + MAX_LABEL_LENGTH + " chars: " + label);
            }
        }
    }

    public Ticket withStage(TicketStage newStage, Instant now) {
        return new Ticket(ticketNo, title, targetRef, clonePath, executorProviderId, executorModel,
                reviewerProviderId, reviewerModel, newStage, createdAt, now,
                execTokenTotal, execTokenSource, agentConfigId, priority, projectId,
                description, note, labels, isSuper);
    }
}

File: gate-web-ui/src/types/ticket.ts
/**
 * 工单 — 对齐后端 ticket 表 (V1__init.sql + V3__cost_metrics.sql + V4 agent_config_id).
 *
 * 字段命名 camelCase 映射列名, 以后端为准 (前端文档 §2.4 B 方案: 手写 types).
 * 后端字段一变须同步本文件.
 */
import type { TicketStage } from './stage';

/** exec_token_source 枚举 (V3): agent_cli | manual | unavailable. */
export type ExecTokenSource = 'agent_cli' | 'manual' | 'unavailable';

export interface Ticket {
  /** ticket_no: 后端全局唯一编号, 如 PROJ-12；项目看板通过 projectId 隔离。 */
  no: string;
  title: string;
  /** 需求描述：会随工单上下文提供给执行侧。 */
  description?: string | null;
  /** 工单备注：面向处理者的补充说明。 */
  note?: string | null;
  /** stage: 见 src/types/stage.ts 状态机. */
  stage: TicketStage;
  /** target_ref: 目标分支 ref (如 refs/heads/main). */
  targetRef: string;
  /** clone_path: 后端为工单创建的独立工作区路径. */
  clonePath?: string;
  /** 最新 review_round — 派生自 presubmit 记录, 无预提审时为 null. */
  reviewRound: number | null;
  /** 最新 tree_hash — 预提审固化锚点 (派生自 presubmit). */
  treeHash: string | null;
  /** 最新 base_commit — 预提审基线 (派生自 presubmit). */
  baseCommit: string | null;
  /** exec_token_total (V3): 执行侧累计 token, 通常 null (agent CLI 未上报). */
  execTokenTotal: number | null;
  /** exec_token_source (V3): agent_cli | manual | unavailable. */
  execTokenSource: ExecTokenSource | null;
  /** agent_config_id (V4): 工单创建时指定的 AgentConfig; null = 系统默认. */
  agentConfigId: string | null;
  createdAt: string;
  updatedAt: string;
  /** 以下为前端展示增强字段（原型阶段本地补充，后端未强制） */
  projectId?: string | null;
  project?: string | null;
  dependencies?: string[];
  labels?: string[];
  priority?: 'P0' | 'P1' | 'P2' | 'P3';
  commentCount?: number;
  branch?: string;
}

/** 建工单请求 (后端文档 §4.2, POST /api/tickets body; description 为 agent 输入). */
export interface CreateTicketRequest {
  /** ticket_no: 后端建工单时使用的唯一编号. */
  ticketNo: string;
  title: string;
  description?: string | null;
  note?: string | null;
  labels?: string[];
  targetRef?: string;
  agentConfigId?: string;
  priority?: Ticket['priority'];
  projectId?: string;
}

export type TicketUpdateRequest = {
  title?: string;
  description?: string | null;
  note?: string | null;
  labels?: string[] | null;
  priority?: Ticket['priority'] | null;
  stage?: TicketStage;
};
